crypto/merkletrie/node.go

```diff
@@ -0,0 +1,383 @@
+		newChildNode.leaf = true
+		newChildNode.hash = d[idiff+1:]
+
+		pnode, nodeID = cache.allocateNewNode()
+		pnode.leaf = false
+		pnode.children = make([]storedNodeIdentifier, 256)
+		pnode.childrenNext = make([]byte, 256)
+
+		pnode.children[n.hash[idiff]] = curChildNodeID
+		pnode.children[d[idiff]] = newChildNodeID
+		if n.hash[idiff] < d[idiff] {
+			pnode.firstChild = n.hash[idiff]
+			pnode.childrenNext[pnode.firstChild] = d[idiff]
+			pnode.childrenNext[d[idiff]] = d[idiff]
+		} else {
+			pnode.firstChild = d[idiff]
+			pnode.childrenNext[pnode.firstChild] = n.hash[idiff]
+			pnode.childrenNext[n.hash[idiff]] = n.hash[idiff]
+		}
+		pnode.hash = append(path, d[:idiff]...)
+
+		for i := idiff - 1; i >= 0; i-- {
+			// create a parent node for pnode.
+			pnode2, nodeID2 := cache.allocateNewNode()
+			pnode2.leaf = false
+			pnode2.children = make([]storedNodeIdentifier, 256)
+			pnode2.childrenNext = make([]byte, 256)
+			pnode2.children[d[i]] = nodeID
+			pnode2.firstChild = d[i]
+			pnode2.childrenNext[d[i]] = d[i]
+			pnode2.hash = append(path, d[:i]...)
+
+			pnode = pnode2
+			nodeID = nodeID2
+		}
+		return nodeID, nil
+	}
+
+	if n.children[d[0]] == storedNodeIdentifierNull {
+		// no such child.
+		var childNode *node
+		var childNodeID storedNodeIdentifier
+		childNode, childNodeID = cache.allocateNewNode()
+		childNode.leaf = true
+		childNode.hash = d[1:]
+
+		pnode, nodeID = n.duplicate(cache)
+		pnode.children[d[0]] = childNodeID
+
+		if pnode.firstChild > d[0] {
+			pnode.childrenNext[d[0]] = pnode.firstChild
+			pnode.firstChild = d[0]
+		} else {
+			// iterate on all the entries.
+			i := pnode.firstChild
+			for {
+				if pnode.childrenNext[i] < d[0] {
+					if pnode.childrenNext[i] == i {
+						pnode.childrenNext[i] = d[0]
+						pnode.childrenNext[d[0]] = d[0]
+						break
+					}
+				} else {
+					pnode.childrenNext[d[0]] = pnode.childrenNext[i]
+					pnode.childrenNext[i] = d[0]
+					break
+				}
+				i = pnode.childrenNext[i]
+			}
+		}
+	} else {
+		// there is already a child there.
+		childNode, err := cache.getNode(n.children[d[0]])
+		if err != nil {
+			return storedNodeIdentifierNull, err
+		}
+		updatedChild, err := childNode.add(cache, d[1:], append(path, d[0]))
+		if err != nil {
+			return storedNodeIdentifierNull, err
+		}
+		pnode, nodeID = n.duplicate(cache)
+		cache.deleteNode(n.children[d[0]])
+		pnode.children[d[0]] = updatedChild
+	}
+	pnode.hash = path
+	return nodeID, nil
+}
+
+// calculateHash calculate the hash of the non-leaf nodes
+// when this function is called, the hashes of all the child node are expected
+// to have been calculated already. This is achived by doing the following:
+// 1. all node id allocations are done in incremental monolitic order, from the bottom up.
+// 2. hash calculations are being doing in node id incremental ordering
+func (n *node) calculateHash(cache *merkleTrieCache) error {
+	if n.leaf {
+		return nil
+	}
+	path := n.hash
```

Comment:
Since `calculateHash` overwrites `n.hash`, this means we can never call `calculateHash` on the same (non-leaf) node twice, or the output will be wrong. Is this the intended behavior?

Reply:
yes. It's not an issue because all the nodes in the trie are immutable. If a non-leaf node need to be modified, a new node would get created to replace it. ( it's also required as part of the update-changes only )